streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java

```diff
@@ -322,6 +323,61 @@ public void testShouldReadFromRegexAndNamedTopics() throws Exception {
     }
 
     @Test
+    public void shouldAlwaysGetNewlyCreatedTopics() throws Exception {
+        final KStreamBuilder builderLeader = new KStreamBuilder();
+        final KStreamBuilder builderFollowerI = new KStreamBuilder();
+        final KStreamBuilder builderFollowerII = new KStreamBuilder();
+
+        final Pattern pattern = Pattern.compile("updating-\\d");
+
+        builderLeader.stream(pattern);
+        builderFollowerI.stream(pattern);
+        builderFollowerII.stream(pattern);
+
+        final KafkaStreams streamsLeader = new KafkaStreams(builderLeader, streamsConfiguration);
+        final KafkaStreams streamsFollowerI = new KafkaStreams(builderFollowerI,
+                                                               streamsConfiguration);
+        final KafkaStreams streamsFollowerII = new KafkaStreams(builderFollowerII,
+                                                                streamsConfiguration);
+
+        final int numberTopicsToAdd = 5;
+
+        try {
+            streamsLeader.start();
+            Thread.sleep(1000);
```

Comment:
I'd suggest avoid `Thread.sleep()` as much as possible in unit test as it will introduce timing-dependent flakiness. If there is any condition that we need to wait on consider using `TestUtils.waitForCondition`.

Also this test seem not always testing the race condition that can cause the issue; I'd suggest make it really "modular" by adding a test for `StreamThread`, in which we do not need to start the thread but just mocking its embedded `StreamPartitionAssignor#onAssignment()` which gives it some assigned partition whose topic is not unknown, and then call its `onPartitionAssigned` callback to check if the topology builder has updated its topology.